jit-compiler/src/compiler.rs

```diff
@@ -0,0 +1,133 @@
+use libc::{c_void, dlopen, dlsym, RTLD_NOW};
+use mktemp::Temp;
+use std::{
+    collections::HashMap,
+    ffi::CString,
+    fs::{self},
+    process::Command,
+    str::from_utf8,
+};
+
+use powdr_ast::{
+    analyzed::Analyzed,
+    parsed::{
+        display::format_type_scheme_around_name,
+        types::{FunctionType, Type, TypeScheme},
+    },
+};
+use powdr_number::FieldElement;
+
+use crate::{codegen::escape_symbol, SymbolMap};
+
+// TODO make this depend on T
+
+const PREAMBLE: &str = r#"
+#![allow(unused_parens)]
+//type FieldElement = powdr_number::GoldilocksField;
+"#;
+
+pub fn generate_glue_code<T: FieldElement>(
+    symbols: &[&str],
+    analyzed: &Analyzed<T>,
+) -> Result<String, String> {
+    let mut glue = String::new();
+    let int_int_fun: TypeScheme = Type::Function(FunctionType {
+        params: vec![Type::Int],
+        value: Box::new(Type::Int),
+    })
+    .into();
+    for sym in symbols {
+        let ty = analyzed.type_of_symbol(sym);
+        if ty != int_int_fun && ty.ty != Type::Col {
+            return Err(format!(
+                "Only (int -> int) functions and columns are supported, but requested{}",
+                format_type_scheme_around_name(sym, &Some(ty)),
+            ));
+        }
+
+        // TODO we should use big int instead of u64
+        let name = escape_symbol(sym);
+        glue.push_str(&format!(
+            r#"
+            #[no_mangle]
+            pub extern fn {}(i: u64) -> u64 {{
+                u64::try_from({name}(ibig::IBig::from(i))).unwrap()
```

Comment:
It would be nice if ibig was dynamically linked both on powdr and on the JITed library, this would ensure that the same runtime version of the library was used on both, reducing the JITed code size and build time, and allowing the type to be used directly on this interface.

Reply:
Which brings another potential issue: since no extern "C" is being used, the ABI is guaranteed to be compatible only if the exact same version who built powdr also builds the library.

Reply:
Yeah, I would not rely on the fact that we have exactly the same rust compiler at runtime as at compile time. Will change to extern c.